File: mvp/src/main/java/com/shea/mvp/activity/BaseActivity.java
package com.shea.mvp.activity;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.shea.mvp.presenter.BasePresenter;

abstract public class BaseActivity<T extends BasePresenter> extends AppCompatActivity {

    protected T presenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutId());
        createPresenter(savedInstanceState);
        presenter.setupViews(savedInstanceState);
    }

    @Override
    protected void onResume() {
        super.onResume();
        presenter.resume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        presenter.pause();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.destroy();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        presenter.onSaveState(outState);
    }

    protected abstract void createPresenter(Bundle restoredBundle);
    protected abstract int getLayoutId();

}

File: src/main/java/com/shea/mvp/activity/BaseActivity.kt
package com.shea.mvp.activity

import android.app.Activity
import android.os.Bundle
import android.support.annotation.IdRes
import android.view.View
import com.shea.mvp.BaseContract
import dagger.android.support.DaggerAppCompatActivity

abstract class BaseActivity<out P : BaseContract.Presenter> : DaggerAppCompatActivity(), BaseContract.View<P> {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layoutId)
        getPresenter().setupViews(savedInstanceState)
        lifecycle.addObserver(getPresenter())
    }

    override final fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        getPresenter().onSaveState(outState)
    }

    fun <T : View> Activity.bind(@IdRes res : Int) : T {
        @Suppress("UNCHECKED_CAST")
        return findViewById(res)
    }
}
